### PR TITLE
Add error handling to the search_logs code path

### PR DESCRIPTION
**Pull request by ekechinwokah**
Adds error handling from the Axum server down to the segment along the search_logs code path. We will add additional error handling to other code paths later.

## What does this PR do?
- Adds error handling to search_logs
 
## How does this PR work? (optional)
- Defined a set of errors in the utils crate so we can return and catch specific errors on failure.

## Refer to a related PR or issue link
- (Related PR or issue)

## Checklist

- [X]  I have written the necessary rustdoc comments.
- [X ]  I have added the necessary unit tests and integration tests. (for non-documentation changes)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,28 +18,24 @@
...
 
 ## Why Infino?
-Telemetry data volumes are increasing exponentially yet there is no  purpose-built storage platform for telemetry. Most observability stacks are built on [ElasticSearch](https://github.com/elastic/elasticsearch-rs), [Clickhouse](https://github.com/ClickHouse/ClickHouse) or [Prometheus](https://github.com/prometheus/prometheus), which are powerful tools but are not built for modern telemetry data so the stacks become complex and expensive to manage. Infino's goal is to reduce the **cost** and **complexity** of observability with smart, high-performance storage for customers and vendors alike.
+Telemetry data volumes are increasing exponentially yet there is no  purpose-built storage platform for telemetry. Most observability stacks are built on [Lucene](https://lucene.apache.org/), [Clickhouse](https://github.com/ClickHouse/ClickHouse) or [Prometheus](https://github.com/prometheus/prometheus), which are powerful tools but are not built for modern telemetry data so the stacks become complex and expensive to manage. Infino's goal is to rethink observability with smart, high-performance telemetry engine for OpenSearch.
```
Comment (vinaykakade) [inline README.md:24]: nit: extra whitespace "no  purpose-built"

File: coredb/src/index_manager/index.rs [hunk 2]
```diff
@@ -594,9 +607,16 @@ mod tests {
     // Prepare an empty JSON body for the query
     let json_body = serde_json::Value::Null;
 
-    // For the query "message", we should expect num_log_messages-1 results.
-    // We collect each message in received_log_messages and then compare it with expected_log_messages.
-    let mut results = index.search_logs("message", json_body.clone(), 0, u64::MAX);
+    // For the query "message", handle errors from search_logs
+    let results = match index.search_logs("message", json_body.clone(), 0, u64::MAX) {
```
Comment (vinaykakade) [inline coredb/src/index_manager/index.rs:611]: You can make this simpler using `if let`. Something like:

let results = if let Ok(results) = index.search_logs("message", json_body.clone(), 0, u64::MAX) {
    results
} else {
    eprintln!("Error in search_logs");
    Vec::new()
};

File: coredb/src/index_manager/index.rs [hunk 3]
```diff
@@ -607,7 +627,13 @@ mod tests {
     assert_eq!(expected_log_messages, received_log_messages);
 
     // For the query "thisisunique", we should expect only 1 result.
-    results = index.search_logs("thisisunique", json_body, 0, u64::MAX);
+    let results = match index.search_logs("thisisunique", json_body, 0, u64::MAX) {
```
Comment (vinaykakade) [inline coredb/src/index_manager/index.rs:630]: Same as above

**Comment by vinaykakade** [comment]
Some minor suggestions. Regarding panicking around locks - there isn't a good way to recover, other than restart the process. I think in that case, a panic is fine. (At some point, we should build a cradle that would restart the server process when it panics.)

**Review by vinaykakade — APPROVED** [review]

File: coredb/src/index_manager/index.rs (hunk 3 above)
Comment (ekechinwokah) [reply coredb/src/index_manager/index.rs:630]: done

File: coredb/src/index_manager/index.rs (hunk 2 above)
Comment (ekechinwokah) [reply coredb/src/index_manager/index.rs:611]: done

File: README.md (hunk 1 above)
Comment (ekechinwokah) [reply README.md:24]: fixed

**Comment by ekechinwokah** [comment]
> Some minor suggestions. Regarding panicking around locks - there isn't a good way to recover, other than restart the process. I think in that case, a panic is fine. (At some point, we should build a cradle that would restart the server process when it panics.)

Ok - perhaps we can leave in the graceful handling for now and change to panics once we have a cradle?

**Comment by vinaykakade** [comment]
> > Some minor suggestions. Regarding panicking around locks - there isn't a good way to recover, other than restart the process. I think in that case, a panic is fine. (At some point, we should build a cradle that would restart the server process when it panics.)
> 
> Ok - perhaps we can leave in the graceful handling for now and change to panics once we have a cradle?

Sure, works

**Pull request merged**